### PR TITLE
BUG: Fix matmul with out= arrays

### PR DESCRIPTION
**Pull request by ev-br**
fixes gh-85

`matmul(x1, x2, out)` does not broadcast x1, x2 against out, like other ufuncs do.

### CODE REVIEW

File: torch_np/tests/numpy_tests/core/test_multiarray.py [hunk 1]
```diff
@@ -5919,9 +5925,10 @@ def test_matrix_matrix_values(self):
         assert_equal(res, tgt12_21)
 
 
-@pytest.mark.xfail(reason='TODO: matmul (ufunc wrapping goes south?)')
 class TestMatmul(MatmulCommon):
-    matmul = np.matmul
+
+    def setup_method(self):
+        self.matmul = np.matmul
```
Comment (ev-br) [inline torch_np/tests/numpy_tests/core/test_multiarray.py:5931]: This change is a bit curious. Storing ufuncs as class attributes seems to work ok in scripts/interactive interpreter, but breaks down in combination with pytest:

the argument processing/normalization machinery picks up the extra `self`-like argument, which leads to 

```
>           tensor = torch.as_tensor(obj)
E           RuntimeError: Could not infer dtype of TestMatmul
```

so there's some spooky action on a distance between meta-stuff pytest is doing and what we do here.

File: torch_np/tests/numpy_tests/core/test_multiarray.py [hunk 2]
```diff
@@ -5974,7 +5981,7 @@ def test_out_contiguous(self):
         # test out non-contiguous
         out = np.ones((5, 2, 2), dtype=float)
         c = self.matmul(a, b, out=out[..., 0])
-        assert c.base is out
+        assert c._tensor._base is out._tensor     # FIXME: self.tensor (no underscore)
```
Comment (ev-br) [inline torch_np/tests/numpy_tests/core/test_multiarray.py:5984]: This will get cleaned up before merging, after the base of the stack gets in.

**Review by lezcano — COMMENTED** [review]
So, I think the broadcasting issue for matmul will simply go away when we implement the `out=` semantics by doing the broadcasting wrt. the out parameter after the computation rather than before. The other matmul fixes from this PR LGTM, so what about simply merging those and revisiting the `out=` behaviour later?

**Comment by ev-br** [comment]
> The other matmul fixes from this PR LGTM, so what about simply merging those and revisiting the out= behaviour later?

Works for me :-). Merging this PR is blocked by the normalizations PR though, as this one is at the top of the stack.

**Comment by ev-br** [comment]
Merged in https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/91

**Pull request closed without merging**